### PR TITLE
ddl: fix bug that range columns partition table add partitions failed

### PR DESCRIPTION
**Pull request by crazycs520**
### What problem does this PR solve? 
```sql
create table t (a datetime) partition by range columns (a) (partition p1 values less than ('2019-06-01'), partition p2 values less than ('2019-07-01'));
alter table t add partition ( partition p3 values less than ('2019-08-01'));  -- should execute succeful.
(1659, u'Field \'"2019-06-01"\' is of a not allowed type for this type of partitioning')
```

### What is changed and how it works?
* Fix Add partition bug.

### Check List 

Tests 

 - Integration test

Code changes

 - Has exported function/method change

Side effects

Related changes

 - Need to cherry-pick to the release branch

Release note

 - Fix the issue that range columns partition table add partitions failed.

### CODE REVIEW

File: ddl/ddl_api.go [hunk 1]
```diff
@@ -2203,7 +2204,11 @@ func (d *ddl) AddTablePartitions(ctx sessionctx.Context, ident ast.Ident, spec *
 	// old partitions to check all partitions is strictly increasing.
 	tmp := *partInfo
 	tmp.Definitions = append(pi.Definitions, tmp.Definitions...)
-	err = checkCreatePartitionValue(ctx, meta, &tmp, t.Cols())
+	if len(tmp.Columns) == 0 {
+		err = checkCreatePartitionValue(ctx, meta, &tmp, t.Cols())
+	} else {
+		err = checkRangeColumnsPartitionValue(ctx, meta, &tmp)
+	}
```
Comment (zimulala) [inline ddl/ddl_api.go]: Do we need to consider `checkPartitionByRange` and `checkPartitionByRangeColumn`?

Comment (crazycs520) [reply ddl/ddl_api.go]: Great, already refactor the code.

File: ddl/ddl_api.go [hunk 2]
```diff
@@ -2185,26 +2175,16 @@ func (d *ddl) AddTablePartitions(ctx sessionctx.Context, ident ast.Ident, spec *
 		return errors.Trace(err)
 	}
 
-	err = checkAddPartitionTooManyPartitions(uint64(len(meta.Partition.Definitions) + len(partInfo.Definitions)))
```
Comment (tiancaiamao) [inline ddl/ddl_api.go]: This check is removed?

File: ddl/ddl_api.go [hunk 3]
```diff
@@ -1563,31 +1555,28 @@ func checkPartitionByRange(ctx sessionctx.Context, tbInfo *model.TableInfo, pi *
...
+	// Check for range columns partition.
 	if err := checkRangeColumnsPartitionType(tbInfo, pi.Columns); err != nil {
 		return err
 	}
 
-	if err := checkRangeColumnsPartitionValue(ctx, tbInfo, pi); err != nil {
-		return err
-	}
-
-	if err := checkNoRangePartitions(len(pi.Definitions)); err != nil {
```
Comment (tiancaiamao) [inline ddl/ddl_api.go]: Why this is removed?

**Comment by tiancaiamao** [comment]
LGTM

**Comment by tiancaiamao** [comment]
@crazycs520 This one is not cherry-pick to release-3.0 !

**Comment by cyliu0** [comment]
@crazycs520 This PR has been labeled `needs-cherry-pick-3.1`. Plz cherry pick this one to release-3.1

**Pull request merged**